tests/unit/test_generate_umm_var.py

```diff
@@ -0,0 +1,276 @@
+            downloadable=True,
+            sort_key='-start_date',
+            concept_id=self.collection_concept_id
+        )
+
+        # Ensure no attempt was made to download the granule after failure:
+        mock_download_granule.assert_not_called()
+
+        # Ensure no attempt was made to publish UMM-Var after the failure:
+        mock_publish_umm_var.assert_not_called()
+
+    @patch('varinfo.umm_var.publish_umm_var')
+    @patch('varinfo.cmr_search.GranuleQuery')
+    @patch('varinfo.generate_umm_var.download_granule')
+    def test_publishing_errors(self, mock_download_granule, mock_granule_query,
+                               mock_publish_umm_var):
+        """ Show the output list from publication will correctly handle a
+            response with an error.
+
+        """
+        error_message = 'Invalid JSON'
+        concept_ids = [
+            'V0000000001-PROV', 'V0000000002-PROV', 'V0000000003-PROV',
+            'V0000000004-PROV', 'V0000000005-PROV', 'V0000000006-PROV',
+            'V0000000007-PROV'
+        ]
+        concept_ids_and_error = concept_ids + [error_message]
+
+        mock_granule_query.return_value.get.return_value = [
+            {
+                'links': [
+                    {
+                        'href': self.netcdf4_url,
+                        'rel': 'http://esipfed.org/ns/fedsearch/1.1/data#'
+                    }
+                ]
+            }
+        ]
+
+        # Add side effect that will copy test file to the temporary directory,
+        # simulating a download.
+        mock_download_granule.side_effect = self.download_granule_side_effect
+
+        mock_publish_umm_var.side_effect = concept_ids_and_error
+
+        # Run the test:
+        published_umm_var = generate_collection_umm_var(
+            self.collection_concept_id,
+            self.bearer_token_header,
+            publish=True
+        )
+
+        # Ensure the granule query used expected query parameters
+        mock_granule_query.return_value.parameters.assert_called_once_with(
+            downloadable=True,
+            sort_key='-start_date',
+            concept_id=self.collection_concept_id
+        )
+
+        # Ensure the call to download the granule had correct parameters
+        mock_download_granule.assert_called_once_with(self.netcdf4_url,
+                                                      self.bearer_token_header,
+                                                      out_directory=ANY)
+
+        # Ensure the output looks as expected
+        # First check all the expected concept IDs are present:
+        self.assertEqual(set(concept_ids),
+                         set(published_umm_var).intersection(concept_ids))
+
+        # Next check the error matches the expected format,
+        # e.g.: '/variable_name: Error Message'.
+        # This is done in pieces, in case dictionary ordering means the
+        # variable associated with the error varies between test runs.
```

Comment:
This is all a bit ugly. It's primarily to guard against the chance that internally the order of variables in variable `VarInfoFromNetCDF4` related dictionaries varies. (Might be no longer an issue with Python ≥ 3.9, but still a bit wary from previous experience with it)

Reply:
Yeah tbh this is a little too obfuscated for me...

Reply:
I followed to here, I think there's only one error message but I don't know what it looks like going in to the stuff below. Meaning I don't see how there's a `/variable_name` on it but I guess I trust you

Reply:
Ok, I'm guessing this is what I hadn't seen yet https://github.com/nasa/earthdata-varinfo/pull/9/files#diff-58979b97c948a78fe40f778dac62d374aba90d16408ddfd23a8829338464bf22R75-R79

Reply:
And I see that @eni-awowale has issues understanding also, maybe concrete examples in the comment?

Reply:
As in beyond:

```
# e.g.: '/variable_name: Error Message'.
```

Or do you mean the comments in the actual code, rather than the tests? (If comments in the code, for sure, will try to improve)

Reply:
- [ ] (I can make that test more commented, too, though)